core/user/user.go

```diff
@@ -67,3 +72,72 @@ func (u UUID) Validate() error {
 func (u UUID) String() string {
 	return string(u)
 }
+
+// userNameTag is the name of the user.
+type userNameTag interface {
+	// Id returns the username string.
+	Id() string
+	// IsLocal indicates if the user is local or external.
+	IsLocal() bool
+}
+
+var (
+	validUserNameSnippet = "[a-zA-Z0-9][a-zA-Z0-9.+-]*[a-zA-Z0-9]"
+	validName            = regexp.MustCompile(fmt.Sprintf("^(?P<name>%s)(?:@(?P<domain>%s))?$", validUserNameSnippet, validUserNameSnippet))
+)
+
+// NewName validates the name and returns a new Name object.
+func NewName(name string) (Name, error) {
+	parts := validName.FindStringSubmatch(name)
+	if len(parts) != 3 {
+		return Name{}, errors.NotValidf("user name %q", name)
+	}
+	domain := parts[2]
+	return Name{
+		name:    name,
+		isLocal: domain == LocalUserDomain || domain == "",
+	}, nil
+}
+
+// Name represents the identity of a user.
+type Name struct {
+	// name is the full username.
+	name string
+	// isLocal indicates if the user is local or external.
+	isLocal bool
```

Comment:
Can we store the domain on Name struct, I suspect if @tlm is right, then we'll want that information.

Reply:
We do have the information contained in the name already, but for consitancy it makes sense to store it in the same way as the user tag.

One thing I'm now a bit unsure about how much we should mirror the tag. The issue is that for the UserTag we have `Name()` and `Id()`. Imo this is needlessly confusing and on `user.Name` there should just be one method `Name()` that returns the full `someone@somewhere`. But this will now mean that they are inconsistent (also that `Name()` doesn't just return `n.name`). I would rather only have one method to get the "name" from the username and I would rather it was not called `Id`. What are your thoughts on this?

I could just remove the `Name` method altogether and just have only `String`. That would remove ambiguity